kong/clustering.lua

```diff
@@ -0,0 +1,409 @@
+local ngx_exit = ngx.exit
+local exiting = ngx.worker.exiting
+local ngx_time = ngx.time
+local new_tab = require("table.new")
+local ngx_var = ngx.var
+local io_open = io.open
+
+
+local MAX_PAYLOAD = 65536 -- 64KB
+local PING_INTERVAL = 30 -- 30 seconds
+local WS_OPTS = {
+  max_payload_len = MAX_PAYLOAD,
+}
+local ngx_ERR = ngx.ERR
+local ngx_DEBUG = ngx.DEBUG
+local ngx_INFO = ngx.INFO
+local WEAK_KEY_MT = { __mode = "k", }
+local CERT_DIGEST
+local CERT, CERT_KEY
+local clients = setmetatable({}, WEAK_KEY_MT)
+local shdict = ngx.shared.kong_clustering -- only when role == "control_plane"
+local prefix = ngx.config.prefix()
+local CONFIG_CACHE = prefix .. "/config.cache.json"
+
+
+local function update_config(config_table, update_cache)
+  assert(type(config_table) == "table")
+
+  local entities, _, _, _, new_hash = dc:parse_table(config_table)
+  if not entities then
+    return nil, "bad config received from control plane"
+  end
+
+  if declarative.get_current_hash() ~= new_hash then
+    -- NOTE: no worker mutex needed as this code can only be
+    -- executed by worker 0
+    local res, err = declarative.load_into_cache_with_events(entities, new_hash)
+    if not res then
+      return nil, err
+    end
+
+    if update_cache then
+      -- local persistence only after load finishes without error
+      local f, err = io_open(CONFIG_CACHE, "w")
+      if not f then
+        ngx_log(ngx_ERR, "unable to open cache file: ", err)
+
+      else
+        local res
+        res, err = f:write(cjson_encode(config_table))
+        if not res then
+          ngx_log(ngx_ERR, "unable to write cache file: ", err)
+        end
+
+        f:close()
+      end
+    end
+
+  else -- get_current_hash() == new_hash
+    ngx_log(ngx_DEBUG, "same config received from control plane,",
+            "no need to reload")
+    return true
+  end
+
+  return true
+end
+
+
+local function send_ping(c)
+  local _, err = c:send_ping(declarative.get_current_hash())
+  if err then
+    ngx_log(ngx_ERR, "unable to ping control plane node: ", err)
+    -- return and let the main thread handle the error
+    return
+  end
+
+  ngx_log(ngx_DEBUG, "sent PING packet to control plane")
+end
+
+
+local function communicate(premature, conf)
+  if premature then
+    -- worker wants to exit
+    return
+  end
+
+  -- TODO: pick one random CP
+  local address = conf.cluster_control_plane
+
+  local c = assert(ws_client:new(WS_OPTS))
+  local uri = "wss://" .. address .. "/v1/outlet?node_id=" ..
+              kong.node.get_id() .. "&node_hostname=" .. utils.get_hostname()
+  local res, err = c:connect(uri, { ssl_verify = true,
+                                    client_cert = CERT,
+                                    client_priv_key = CERT_KEY,
+                                    server_name = "kong_clustering",
+                                  }
+                            )
+  if not res then
+    local delay = 9 + math.random()
```

Comment:
what is the meaning of `math.random` here?

Reply:
This aims to evenly distribute reconnect requests so that the CP don't get overwhelmed by lots of client connecting at once.

Reply:
Could we consider increasing this jitter spread? Right now this is just +/- half a second. Spreading this among 5-10 seconds seems more reasonable.

Reply:
How does this look: https://github.com/Kong/kong/pull/5294/commits/da771e710bc6171c05f40407c04214c599447683

Reply:
Seems very reasonable. In the future an exponential backoff approach might be more valuable but i think this is an excellent step as-is :+1: